rust/src/pgsql/parser.rs

```diff
@@ -0,0 +1,2089 @@
+use nom::character::streaming::alphanumeric1;
+use nom::number::streaming::{be_u8, be_u16, be_u32};
+use nom::number::streaming::{be_i16, be_i32};
+
+pub const PGSQL_DUMMY_PROTO_MAJOR: u16 = 1234; // 0x04d2
+pub const PGSQL_DUMMY_PROTO_MINOR_SSL: u16 = 5679; //0x162f
+pub const _PGSQL_DUMMY_PROTO_MINOR_GSSAPI: u16 = 5680; // 0x1630
+
+#[derive(Debug, PartialEq)]
+pub struct PgsqlParameter {
+    pub name: Vec<u8>,
+    pub value: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PgsqlStartupParameters {
+    pub user: PgsqlParameter,
+    pub database: Option<PgsqlParameter>,
+    // TODO get rid of this Option<vec>
+    pub optional_params: Option<Vec<PgsqlParameter>>,
+}
+
+impl fmt::Display for PgsqlStartupParameters {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DummyStartupPacket {
+    length: u32,
+    proto_major: u16,
+    proto_minor: u16,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct StartupPacket{
+    pub length: u32,
+    pub proto_major: u16,
+    pub proto_minor: u16,
+    pub params: PgsqlStartupParameters,
+}
+
+impl fmt::Display for StartupPacket {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct RegularPacket{
+    pub identifier: u8,
+    pub length: u32,
+    pub payload: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PgsqlErrorNoticeMessageField{
+    pub field_type: PgsqlErrorNoticeFieldType,
+    pub field_value: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct ErrorNoticeMessage {
+    pub identifier: u8,
+    pub length: u32,
+    pub message_body: Vec<PgsqlErrorNoticeMessageField>,
+}
+
+impl ErrorNoticeMessage {
+    pub fn new(identifier: u8, length: u32) -> Self {
+        ErrorNoticeMessage{
+            identifier,
+            length,
+            message_body: Vec::new(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum SSLResponseMessage {
+    SSLAccepted,
+    SSLRejected,
+    InvalidResponse,
+}
+
+impl From<u8> for SSLResponseMessage {
+    fn from(identifier: u8) -> Self {
+        match identifier {
+            b'S' => Self::SSLAccepted,
+            b'N' => Self::SSLRejected,
+            _ => Self::InvalidResponse,
+        }
+    }
+}
+
+impl From<char> for SSLResponseMessage {
+    fn from(identifier: char) -> Self {
+        match identifier {
+            'S' => Self::SSLAccepted,
+            'N'=> Self::SSLRejected,
+            _ => Self::InvalidResponse,
+        }
+    }
+}
+
+// TODO - do I really need this?
+impl fmt::Display for SSLResponseMessage {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            SSLResponseMessage::SSLAccepted => write!(f, "SSL Accepted"),
+            SSLResponseMessage::SSLRejected => write!(f, "SSL Rejected"),
+            _ => write!(f, "Invalid server response"),
+        }
+    }
+}
```

Comment:
No, I don't think so.  You'd have to match on the variants in your logger without this though.  But thats not a bad thing.  This invokes the formatter which is slow, and creates a string, just to have it dereferenced in the logger.  So may as well just match in the logger and use static strings.

Or even better, create an `impl` for `SSLResponseMessage` and add a `to_str()` method that returns the pretty string as a static.

Reply:
Thanks, I shall do the same for the other `fmt::Display` that are there, then, too.